### PR TITLE
release-23.2: logmetrics: add metrics for fluent write error, attempts

### LINKED ISSUE #116694: Improve observability around fluent logging sinks
**Is your feature request related to a problem? Please describe.**

Although we're adding the log.fluent.sink.conn.errors metric in 23.2 to track fluent sink connection errors, there are a few pieces of information that would also be useful for monitoring the flow of logs from crdb. Specifically, it would be helpful to track write errors in addition to connection errors (since we've seen log lines indicating write errors but not connection errors), and to track attempts as well as errors, so that we can calculate error rates.

**Describe the solution you'd like**

Add a new metric for fluent sink write errors, as well as metrics for connect and write attempts.

**Describe alternatives you've considered**

In theory users could alert on log lines that indicate errors sending logs to the fluent sink, but obviously this could be tricky if logs aren't flowing to the system that would alert on them!

**Additional context**

This looks like it would be pretty straightforward to implement, so I'm going to send in a tiny patch for discussion.

Jira issue: CRDB-34726

### PR DESCRIPTION
**Pull request by jmcarp**
Backport 1/1 commits from #116699.

/cc @cockroachdb/release

---

Previously, we emitted a metric to count fluent sink connection errors in the logmetrics package. However, we didn't have a metric to count write errors, and we didn't have metrics to count connection/write attempts. This metrics would let users write better alerts. Specifically, users would be able to alert on write errors as well as connection errors. And users would be able to alert on error rates rather than error counts. This patch adds new metrics to track fluent sink connection attempts, and fluent sink write attempts and errors.

Fixes #116694.

Release justification: low-risk and necessary for monitoring log delivery in 23.2.

### CODE REVIEW

**Review by dhartunian — COMMENTED** [review]
Backporting new functionality requires a conditional check. Is there a straightforward way we could gate these new metrics behind an env var or something?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117325#-:-Nndq0fHCpATtIPqNz9y:b3n93er)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @joshimhoff)

**Review by joshimhoff — COMMENTED** [review]
LGTM from the perspective of the backport policy. One Q. Re: idea of putting this behind an env var, whatever David thinks sounds good to me.

File: pkg/util/log/fluent_client.go [hunk 1]
```diff
@@ -67,17 +67,24 @@ func (l *fluentSink) exitCode() exit.Code {
...
 		return nil
 	}
 
+	logging.metrics.IncrementCounter(FluentSinkConnectionAttempt, 1)
 	if err := l.ensureConnLocked(b); err != nil {
 		logging.metrics.IncrementCounter(FluentSinkConnectionError, 1)
 		return err
 	}
-	return l.tryWriteLocked(b)
+	if err := l.tryWriteLocked(b); err != nil {
```
[the remark this answers is not in the document]
Comment (jmcarp) [reply pkg/util/log/fluent_client.go:83]: I'm going to defer to @abarganier on this, but I think the metrics logic here is correct. Here's how I understand the code to work:

* We try to write to the sink with `tryWriteLocked`. If this method succeeds, we're done; else we set `l.mu.good` to `false`.
* If `l.mu.good` is `false`, we attempt to reconnect to the tcp sink.
* We try again to write to the sink.

We might call `tryWriteLocked` twice, but that reflects the fact that the first call is expected to fail if we've lost our connection to the tcp sink. So I think it's correct to increment the write attempt counter just once for the `output` method.

**Review by dhartunian — APPROVED** [review]
Given the discussion in Slack and no blockers from leadership we can move ahead with this as-is without an env var since it's low risk and high-value: https://cockroachlabs.slack.com/archives/C060S8QKABT/p1704729956892999

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117325#-:-Np6OYwf9774q6Hqcj2O:b-mrftxh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @abarganier)

File: pkg/util/log/fluent_client.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (joshimhoff) [reply pkg/util/log/fluent_client.go:83]: Ack!

**Review by joshimhoff — APPROVED** [review]
LGTM too.

**Pull request merged**